packages/css-parser-algorithms/src/consume/consume-component-block-function.ts

```diff
@@ -0,0 +1,621 @@
+import { ComponentValueType } from '../util/component-value-type';
+
+export type ContainerNode = FunctionNode | SimpleBlockNode;
+
+export type ComponentValue = FunctionNode | SimpleBlockNode | WhitespaceNode | CommentNode | TokenNode | UnclosedSimpleBlockNode | UnclosedFunctionNode;
+
+// https://www.w3.org/TR/css-syntax-3/#consume-a-component-value
+export function consumeComponentValue(ctx: Context, tokens: Array<CSSToken>): { advance: number, node: ComponentValue } {
+	const token = tokens[0];
+	if (
+		token[0] === TokenType.OpenParen ||
+		token[0] === TokenType.OpenCurly ||
+		token[0] === TokenType.OpenSquare
+	) {
+		const r = consumeSimpleBlock(ctx, tokens);
+		return {
+			advance: r.advance,
+			node: r.node,
+		};
+	}
+
+	if (token[0] === TokenType.Function) {
+		const r = consumeFunction(ctx, tokens);
+		return {
+			advance: r.advance,
+			node: r.node,
+		};
+	}
+
+	if (token[0] === TokenType.Whitespace) {
+		const r = consumeWhitespace(ctx, tokens);
+		return {
+			advance: r.advance,
+			node: r.node,
+		};
+	}
+
+	if (token[0] === TokenType.Comment) {
+		const r = consumeComment(ctx, tokens);
+		return {
+			advance: r.advance,
+			node: r.node,
+		};
+	}
+
+	return {
+		advance: 1,
+		node: new TokenNode(token),
+	};
+}
+
+export class FunctionNode {
+	type: ComponentValueType = ComponentValueType.Function;
+
+	name: TokenFunction;
+	endToken: CSSToken;
+	value: Array<ComponentValue>;
+
+	constructor(name: TokenFunction, endToken: CSSToken, value: Array<ComponentValue>) {
+		this.name = name;
+		this.endToken = endToken;
+		this.value = value;
+	}
+
+	get nameTokenValue(): string {
+		return this.name[4].value;
+	}
+
+	tokens(): Array<CSSToken> {
+		return [
+			this.name,
+			...this.value.flatMap((x) => {
+				if (isToken(x)) {
+					return x;
+				}
+
+				return x.tokens();
+			}),
+			this.endToken,
+		];
+	}
+
+	toString(): string {
+		const valueString = this.value.map((x) => {
+			if (isToken(x)) {
+				return stringify(x);
+			}
+
+			return x.toString();
+		}).join('');
+
+		return stringify(this.name) + valueString + stringify(this.endToken);
+	}
+
+	indexOf(item: ComponentValue): number | string {
+		return this.value.indexOf(item);
+	}
+
+	at(index: number | string) {
+		if (typeof index === 'number') {
+			if (index < 0) {
+				index = this.value.length + index;
+			}
+			return this.value[index];
+		}
+	}
+
+	walk(cb: (entry: { node: ComponentValue, parent: ContainerNode }, index: number | string) => boolean | void) {
+		let aborted = false;
+
+		this.value.forEach((child, index) => {
+			if (aborted) {
+				return;
+			}
+
+			if (cb({ node: child, parent: this }, index) === false) {
+				aborted = true;
+				return;
+			}
+
+			if ('walk' in child) {
+				if (child.walk(cb) === false) {
+					aborted = true;
+					return;
+				}
+			}
+		});
+
+		if (aborted) {
+			return false;
+		}
+	}
+
+	toJSON() {
+		return {
+			type: this.type,
+			name: this.name[4].value,
```

Comment:
Perhaps we could/should reuse `nameTokenValue` getter here?

Reply:
yes, and `nameTokenValue` should not be a getter.
I think it is the only getter in all the three new packages.

Reply:
updated